### PR TITLE
SnipSync prototype

### PR DESCRIPTION
**Pull request by cullywakelin**
This PR adds a working prototype to the repository.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Dependencies
```
Comment (rylandg) [inline .gitignore:1]: Might be better to use a boilerplate gitignore. For example .DS_file is not ignored.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
+# SnipSync
```
Comment (rylandg) [inline README.md:1]: Snipsync (no caps)

File: src/common.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  rootDir: process.cwd(),
+  cfgFile: 'snipsync.config.yaml',
+  extractionDir: 'sync_repos',
+  markdownCodeTicks: '```',
+  fmtStartCodeBlock: (ext) => '```' + ext,
+  readStart: '@@@START',
+  readEnd: '@@@END',
+  writeStart: '<!--START',
+  writeEnd: '<!--END',
+  fmtProgressBar: (message) => '⭐ ' + `| {bar} | {percentage}% | {value}/{total} chunks | ${message}`
```
Comment (rylandg) [inline src/common.js:11]:
```suggestion
  fmtProgressBar: (message) => `⭐   | {bar} | {percentage}% | {value}/{total} chunks | ${message}`
```

File: src/config.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+const { join } = require('path');
+const { sync } = require('node-read-yaml');
+const { cfgFile, rootDir, fmtProgressBar } = require('./common.js');
```
Comment (rylandg) [inline src/config.js:3]: You don't need trailing js extension.

```suggestion
const { cfgFile, rootDir, fmtProgressBar } = require('./common');
```

File: src/config.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+const progress = require ('cli-progress');
+
+function readConfig() {
```
Comment (rylandg) [inline src/config.js:6]:
```suggestion
module.exports.readConfig = () => {
```

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+const logger = require('js-logger');
+const config = require ('./config.js');
+const sync = require ('./sync.js');
+const common = require ('./common.js');
+
+logger.useDefaults();
+
+cfg = config.readconfig();
+
+synctron = new sync.Sync(cfg, logger);
```
Comment (rylandg) [inline src/index.js:11]: Always use var declaration types

```suggestion
const cfg = config.readconfig();

const synctron = new sync.Sync(cfg, logger);
```

File: src/snippet.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+const {
+  fmtStartCodeBlock,
+  markdownCodeTicks
```
Comment (rylandg) [inline src/snippet.js:3]: Leave a trailing comma on objects
```suggestion
  markdownCodeTicks,
```

Comment (rylandg) [reply src/snippet.js:3]: Also if a file exports a class the filename should be cased. Ie: Snippet.js

File: src/sync.js [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+const { eachLine } = require('line-reader');
+const {
+  extractionDir,
+  fmtProgressBar,
+  markdownCodeTicks,
+  readStart,
+  readEnd,
+  rootDir,
+  writeStart,
+  writeEnd
```
Comment (rylandg) [inline src/sync.js:13]: Trailing comma please

File: src/sync.js [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+  async writeFiles(files) {
+    let insertRootPath = join(rootDir, this.config.target);
+    let writeFileProgress = new progress.Bar({
+      format: fmtProgressBar('writing files to ' + insertRootPath),
+      barsize: 20
+    }, progress.Presets.shades_classic);
+    writeFileProgress.start(files.length, 0);
+    for (let i = 0; i< files.length; i++) {
+      let file = files[i];
+      let fileString = file.lines.join('\n');
```
Comment (rylandg) [inline src/sync.js:272]:
```suggestion
      const fileString = `${file.lines.join('\n')}\n;
```

File: src/sync.js [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+
+  async writeFiles(files) {
+    let insertRootPath = join(rootDir, this.config.target);
+    let writeFileProgress = new progress.Bar({
+      format: fmtProgressBar('writing files to ' + insertRootPath),
+      barsize: 20
+    }, progress.Presets.shades_classic);
+    writeFileProgress.start(files.length, 0);
+    for (let i = 0; i< files.length; i++) {
+      let file = files[i];
```
Comment (rylandg) [inline src/sync.js:271]:
```suggestion
      const file = files[i];
```

File: src/sync.js [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+
+  async spliceFile(start, end, snippet, file) {
+    let rmlines = end - start;
+    file.lines.splice(start, rmlines - 1, ...snippet.lines);
+    return file;
+  }
+
+  async writeFiles(files) {
+    let insertRootPath = join(rootDir, this.config.target);
+    let writeFileProgress = new progress.Bar({
```
Comment (rylandg) [inline src/sync.js:265]:
```suggestion
    const writeFileProgress = new progress.Bar({
```

File: src/sync.js [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+  }
+
+  async spliceFile(start, end, snippet, file) {
+    let rmlines = end - start;
+    file.lines.splice(start, rmlines - 1, ...snippet.lines);
+    return file;
+  }
+
+  async writeFiles(files) {
+    let insertRootPath = join(rootDir, this.config.target);
```
Comment (rylandg) [inline src/sync.js:264]:
```suggestion
    const insertRootPath = join(rootDir, this.config.target);
```

File: src/sync.js [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+    let cleanupProgress = new progress.Bar({
+      format: fmtProgressBar('cleaning up downloads'),
+      barsize: 20
+    }, progress.Presets.shades_classic);
+    cleanupProgress.start(1, 0);
+    let path = join(rootDir, extractionDir);
+    rimrafAsync(path);
+    cleanupProgress.update(1);
+    cleanupProgress.stop();
+    return;
```
Comment (rylandg) [inline src/sync.js:292]: No need for this return

**Review by rylandg — CHANGES_REQUESTED** [review]
Overall much improved. Many things I only commented once even though its an issue across multiple files. Try and propagate where it makes sense.

File: src/snippet.js (hunk 7 above)
Comment (cullywakelin) [reply src/snippet.js:3]: I don't see this happening with any other node modules?

Comment (cullywakelin) [reply src/snippet.js:3]: OK nvm I found other articles that mention this as well.

**Review by rylandg — APPROVED** [review]
Good enough for first version. Will need to add linter/typescript.

**Pull request merged**